File: src/lib/tauri/sales.ts
import { nativeRequest } from "./client";
import type {
  CompleteSaleInput,
  CompleteSaleResult,
  PosProductPage,
  PosQuery,
  ReturnSaleInput,
  Sale,
  SaleDetail,
  SalePage,
  SaleQuery,
  SaleReturn,
  VoidSaleInput,
} from "./sales.types";
export * from "./sales.types";
export const completeSale = (input: CompleteSaleInput) =>
  nativeRequest<CompleteSaleResult>("complete_sale", { input });
export const returnSale = (input: ReturnSaleInput) =>
  nativeRequest<SaleReturn>("return_sale", { input });
export const voidSale = (input: VoidSaleInput) =>
  nativeRequest<Sale>("void_sale", { input });
export const listSales = (query: SaleQuery) =>
  nativeRequest<SalePage>("list_sales", { query });
export const getSale = (id: string) => nativeRequest<SaleDetail>("get_sale", { id });
export const posSearch = (query: PosQuery) =>
  nativeRequest<PosProductPage>("pos_search", { query });

File: src-tauri/src/commands/sales.rs
use crate::{
    commands::branches::DatabaseState,
    db::{
        models::sales::*,
        repositories::{sales},
        services,
    },
    error::AppError,
};
use tauri::State;
/// Thin POS commands. Business rules live in db::services; queries in
/// db::repositories. All amounts are integer EGP minor units.

#[tauri::command]
pub fn complete_sale(
    state: State<'_, DatabaseState>,
    input: CompleteSaleInput,
) -> Result<CompleteSaleResult, AppError> {
    services::sales::complete_sale(state.db()?, &input)
}
#[tauri::command]
pub fn return_sale(
    state: State<'_, DatabaseState>,
    input: ReturnSaleInput,
) -> Result<SaleReturn, AppError> {
    services::sales::return_sale(state.db()?, &input)
}
#[tauri::command]
pub fn void_sale(
    state: State<'_, DatabaseState>,
    input: VoidSaleInput,
) -> Result<Sale, AppError> {
    services::sales::void_sale(state.db()?, &input)
}
#[tauri::command]
pub fn list_sales(
    state: State<'_, DatabaseState>,
    query: SaleQuery,
) -> Result<SalePage, AppError> {
    sales::list(state.db()?, &query)
}
#[tauri::command]
pub fn get_sale(
    state: State<'_, DatabaseState>,
    id: String,
) -> Result<SaleDetail, AppError> {
    sales::detail(state.db()?, &id)
}
#[tauri::command]
pub fn pos_search(
    state: State<'_, DatabaseState>,
    query: PosQuery,
) -> Result<PosProductPage, AppError> {
    sales::pos_search(state.db()?, &query)
}